_posts/2019-10-16-deployment-updates.md

```diff
@@ -0,0 +1,214 @@
+---
+layout: post
+title: Dask Deployment Updates
+tags: []
+theme: twitter
+---
+{% include JB/setup %}
+
+Summary
+-------
+
+Over the last six months many Dask developers have worked on making Dask easier
+to deploy in a wide variety of situations.  This post summarizes those
+efforts, and provides links to ongoing work.
+
+What we mean by Deployment
+--------------------------
+
+In order to run Dask on a cluster, you need to setup a scheduler on one
+machine:
+
+```
+$ dask-scheduler
+Scheduler running at tcp://192.168.0.1
+```
+
+And start Dask workers on many other machines
+
+```
+$ dask-worker tcp://192.168.0.1
+$ dask-worker tcp://192.168.0.1
+$ dask-worker tcp://192.168.0.1
+$ dask-worker tcp://192.168.0.1
+```
+
+For informal clusters people might do this manually, by logging into each
+machine and running these commands themselves.  However it's much more common
+to use a cluster resource manager such as Kubernetes, Yarn (Hadoop/Spark),
+SGE, PBS, SLURM, LSF, some cloud service or some custom system.
+
+As Dask is used by more institutions and used more broadly within those
+institutions making deployment smooth and natural becomes increasingly
+important.  This is so important in fact, that there have been seven separate
+efforts to improve deployment in some regard or another by a few different
+groups.
+
+We'll briefly summarize and link to this work below, and then we'll finish up
+by talking about some internal design that helped to make this work more
+consistent.
+
+
+Dask-SSH
+--------
+
+According to our user survey, the most common deployment mechanism was still
+SSH.  Dask has had a [command line dask-ssh
+tool](https://docs.dask.org/en/latest/setup/ssh.html#command-line) to make it
+easier to deploy with SSH for some time.  We recently updated this to also
+include a Python interface, which provides more control.
+
+```python
+>>> from dask.distributed import Client, SSHCluster
+>>> cluster = SSHCluster(
+...     ["host1", "host2", "host3", "host4"],
+...     connect_options={"known_hosts": None},
+...     worker_options={"nthreads": 2},
+...     scheduler_options={"port": 0, "dashboard_address": ":8797"}
+... )
+>>> client = Client(cluster)
+```
+
+This isn't what we recommend for large institutions, but it can be helpful for
+more informal groups who are just getting started.
+
+
+Dask-Jobqueue and Dask-Kubernetes Rewrite
+-----------------------------------------
+
+We've rewritten Dask-Jobqueue for SLURM/PBS/LSF/SGE cluster managers typically
```

Comment:
I read through the whole thing. This seems in a good state to me already.

Maybe something that could be explicitly mentioned is the version of Dask-Jobqueue (0.7.0) and Dask-Kubernetes (0.10.0 I think, please someone double-check) that benefit from this rewrite.

Reply:
Not sure whether we should link to the associated PRs, e.g. https://github.com/dask/dask-jobqueue/pull/307 for `dask-jobqueue` or stay high-level. I have to say the Dask-Jobqueue PR definitely has all of the details but a lot of noise and I don't think it would be that useful to add the PR link.

Reply:
Yep 0.10.0